### PR TITLE
Сорокин Андрей. Задача 1. Вариант 5. Технология SEQ. Умножение разреженных матриц. Элементы типа double. Формат хранения матрицы – столбцовый (CCS).

### PR DESCRIPTION
**Pull request by AndreySorokin7**
Постановка задачи
Задача заключается в реализации последовательного алгоритма умножения разреженных матриц, представленных в формате CCS (Compressed Column Storage). Алгоритм принимает на вход две матрицы в формате CCS и вычисляет их произведение, сохраняя результат также в CCS-формате.

Представление данных:
Матрицы хранятся в разреженном формате CCS, включающем:
values — массив ненулевых элементов.
row_indices — массив индексов строк для ненулевых элементов.
col_ptr — массив указателей на начало каждого столбца.
Входные матрицы:
A размерности M×K.
B размерности K×N.
Выходная матрица 
C размерности M×N сохраняется в том же формате CCS.

Этапы алгоритма:
PreProcessingImpl(): Инициализация входных и выходных данных.
ValidationImpl(): Проверка корректности входных данных.
RunImpl(): Основное вычисление произведения матриц.
PostProcessingImpl(): Запись результата в выходной буфер.

### CODE REVIEW

File: tasks/seq/sorokin_a_multiplication_sparse_matrices_double_ccs/func_tests/main.cpp [hunk 1]
```diff
@@ -0,0 +1,284 @@
+#include <gtest/gtest.h>
+
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+#include "core/task/include/task.hpp"
+#include "seq/sorokin_a_multiplication_sparse_matrices_double_ccs/include/ops_seq.hpp"
+
```
Comment (aobolensk) [inline tasks/seq/sorokin_a_multiplication_sparse_matrices_double_ccs/func_tests/main.cpp:10]: Please, add corner cases tests (empty matrices, incompatible matrices, etc.). Also compare CCS matrix multiplication with naive matrix multiplication (with regular matrices, 3 for-loops) on relatively small matrix sizes in the tests

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: tasks/seq/sorokin_a_multiplication_sparse_matrices_double_ccs/func_tests/main.cpp (hunk 1 above)
Comment (AndreySorokin7) [reply tasks/seq/sorokin_a_multiplication_sparse_matrices_double_ccs/func_tests/main.cpp:10]: I'm sorry, but it's not clear enough to me what I need to do. The description contains information that three parameters m k n are used. These parameters store the sizes of the transmitted matrices. The check that they are not equal to 0 (not empty) is in validation. And even special tests have been written for it. Three parameters are passed, and even in theory it is impossible for disproportionate multiplication matrices to be passed. Therefore, there are tests for extreme cases, and checking each of the three arrays looks redundant. Regarding the naive multiplication with 3 for. What is this for and in what format? It's just that everything is stored in ccs at once and I need to come up with something and write more algorithms to somehow translate it and write them somewhere separately in their tests. If this is necessary for correctness and is needed on small matrices, I can better use comments in the code to show what the matrix is in its usual form and what should happen. I think that in this case it will be even easier to verify than writing all these translations.

File: tasks/seq/sorokin_a_multiplication_sparse_matrices_double_ccs/src/ops_seq.cpp [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+#include <algorithm>
+#include <cmath>
+#include <cstddef>
+#include <vector>
+
+namespace sorokin_a_multiplication_sparse_matrices_double_ccs_seq {
+void MultiplyCCS(const std::vector<double> &a_values, const std::vector<int> &a_row_indices, int m,
+                 const std::vector<int> &a_col_ptr, const std::vector<double> &b_values,
+                 const std::vector<int> &b_row_indices, int k, const std::vector<int> &b_col_ptr,
+                 std::vector<double> &c_values, std::vector<int> &c_row_indices, int n, std::vector<int> &c_col_ptr) {
```
Comment (Semyon1104) [inline tasks/seq/sorokin_a_multiplication_sparse_matrices_double_ccs/src/ops_seq.cpp:12]: why are variables passed to it in such a strange order?

File: tasks/seq/sorokin_a_multiplication_sparse_matrices_double_ccs/func_tests/main.cpp [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+TEST(sorokin_a_multiplication_sparse_matrices_double_ccs_seq, test_1) {
```
Comment (Semyon1104) [inline tasks/seq/sorokin_a_multiplication_sparse_matrices_double_ccs/func_tests/main.cpp:11]: please change test's names

File: tasks/seq/sorokin_a_multiplication_sparse_matrices_double_ccs/src/ops_seq.cpp [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  C_values_ = std::vector<double>(output_size, 0);
+  C_row_indices_ = std::vector<int>(output_size, 0);
+  C_col_ptr_ = std::vector<int>(output_size, 0);
+  return true;
+}
+
+bool sorokin_a_multiplication_sparse_matrices_double_ccs_seq::TestTaskSequential::ValidationImpl() {
+  // Check equality of counts elements
+  return task_data->inputs_count[0] != 0 && task_data->inputs_count[1] != 0 && task_data->inputs_count[2] != 0;
+}
```
Comment (Semyon1104) [inline tasks/seq/sorokin_a_multiplication_sparse_matrices_double_ccs/src/ops_seq.cpp:87]: few checks for your task

File: tasks/seq/sorokin_a_multiplication_sparse_matrices_double_ccs/func_tests/main.cpp (hunk 3 above)
Comment (AndreySorokin7) [reply tasks/seq/sorokin_a_multiplication_sparse_matrices_double_ccs/func_tests/main.cpp]: fixed

File: tasks/seq/sorokin_a_multiplication_sparse_matrices_double_ccs/src/ops_seq.cpp (hunk 2 above)
Comment (AndreySorokin7) [reply tasks/seq/sorokin_a_multiplication_sparse_matrices_double_ccs/src/ops_seq.cpp:12]: The tidy bot requires that variables with the same data types should not be in a row, so they had to be placed so that they were, as it were, some separators.

**Review by Semyon1104 — APPROVED** [review]

File: tasks/seq/sorokin_a_multiplication_sparse_matrices_double_ccs/func_tests/main.cpp (hunk 1 above)
Comment (aobolensk) [reply tasks/seq/sorokin_a_multiplication_sparse_matrices_double_ccs/func_tests/main.cpp:10]: > The check that they are not equal to 0 (not empty) is in validation. 

If they don't pass validation, then just assert that validation return false

> And even special tests have been written for it

Yes, but there are quite a few of them

> Regarding the naive multiplication with 3 for. ... I can better use comments in the code to show what the matrix is in its usual form and what should happen. 

This is also acceptable

Comment (AndreySorokin7) [reply tasks/seq/sorokin_a_multiplication_sparse_matrices_double_ccs/func_tests/main.cpp:10]: image
Here is a validation test where an empty matrix comes in. Also, there is a test where another matrix is empty. It doesn't make sense to add more tests for this, since in validation all cases when they are empty are considered.

Comment (AndreySorokin7) [reply tasks/seq/sorokin_a_multiplication_sparse_matrices_double_ccs/func_tests/main.cpp:10]: I added comments with matrices to the functional tests so that, if necessary, I could pre-calculate

**Review by ermolaevv — APPROVED** [review]

**Review by aobolensk — APPROVED** [review]

**Review by allnes — APPROVED** [review]
